x/concentrated-liquidity/math.go

```diff
@@ -76,36 +98,66 @@ func calcAmount1Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec
+	amountIn = calcAmount0Delta(liquidity, sqrtPriceTarget, sqrtPriceCurrent, true)
 	if amountRemaining.GTE(amountIn) {
 		sqrtPriceNext = sqrtPriceTarget
 	} else {
-		sqrtPriceNext = getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining, zeroForOne)
+		sqrtPriceNext = s.getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining)
 	}
+	amountIn = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, true)
+	amountOut = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
 
-	if zeroForOne {
-		amountIn = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, true)
-		amountOut = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
+	return sqrtPriceNext, amountIn, amountOut
+}
+
+func (s *oneForZeroStrategy) computeSwapStep(sqrtPriceCurrent, sqrtPriceTarget, liquidity, amountRemaining sdk.Dec) (sqrtPriceNext, amountIn, amountOut sdk.Dec) {
+	amountIn = calcAmount1Delta(liquidity, sqrtPriceTarget, sqrtPriceCurrent, true)
+	if amountRemaining.GTE(amountIn) {
+		sqrtPriceNext = sqrtPriceTarget
 	} else {
-		amountIn = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, true)
-		amountOut = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
+		sqrtPriceNext = s.getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining)
 	}
+	amountIn = calcAmount1Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, true)
+	amountOut = calcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
 
 	return sqrtPriceNext, amountIn, amountOut
 }
 
-func getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining sdk.Dec, zeroForOne bool) (sqrtPriceNext sdk.Dec) {
-	if zeroForOne {
-		sqrtPriceNext = getNextSqrtPriceFromAmount0RoundingUp(sqrtPriceCurrent, liquidity, amountRemaining)
-	} else {
-		sqrtPriceNext = getNextSqrtPriceFromAmount1RoundingDown(sqrtPriceCurrent, liquidity, amountRemaining)
+func (s *zeroForOneStrategy) getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining sdk.Dec) (sqrtPriceNext sdk.Dec) {
+	return getNextSqrtPriceFromAmount0RoundingUp(sqrtPriceCurrent, liquidity, amountRemaining)
+}
+
+func (s *oneForZeroStrategy) getNextSqrtPriceFromInput(sqrtPriceCurrent, liquidity, amountRemaining sdk.Dec) (sqrtPriceNext sdk.Dec) {
+	return getNextSqrtPriceFromAmount1RoundingDown(sqrtPriceCurrent, liquidity, amountRemaining)
+}
+
+func (s *zeroForOneStrategy) setLiquidityDeltaSign(deltaLiquidity sdk.Int, swapIn bool) sdk.Int {
+	if swapIn {
+		return deltaLiquidity.Neg()
 	}
-	return sqrtPriceNext
+	return deltaLiquidity
+}
+
+func (s *oneForZeroStrategy) setLiquidityDeltaSign(deltaLiquidity sdk.Int, swapIn bool) sdk.Int {
+	if !swapIn {
+		return deltaLiquidity.Neg()
+	}
+	return deltaLiquidity
+}
+
+func (s *zeroForOneStrategy) setNextTick(nextTick int64, swapIn bool) sdk.Int {
+	if swapIn {
+		return sdk.NewInt(nextTick - 1)
+	}
+	return sdk.NewInt(nextTick)
```

Comment:
Why do we need to pass in the `swapIn` flag? Can we execute one or the other depending on the strategy?

The point of using the strategy pattern is to be able to avoid passing in these flags

Reply:
Yes I see, thanks you